File: Fundamentals_of_Software_Design/1_Basic_skills/Счастливые билеты (продвинутая версия)/main.py
import random

def is_lucky_ticket(ticket):
    sum1 = sum(int(ticket[i]) for i in range(16))
    sum2 = sum(int(ticket[i]) for i in range(16, 32))
    return sum1 == sum2

def generate_ticket():
    # Генерируем случайный 32-значный номер билета
    return ''.join(str(random.randint(0, 9)) for _ in range(32))

def main():
    total_tickets = 1000000  # Количество билетов для проверки
    lucky_count = 0  # Количество счастливых билетов

    for _ in range(total_tickets):
        ticket = generate_ticket()
        if is_lucky_ticket(ticket):
            lucky_count += 1

    # Вычисляем долю счастливых билетов
    lucky_percentage = (lucky_count / total_tickets) * 100

    print(f"Доля счастливых билетов: {lucky_percentage:.6f}%")

if __name__ == "__main__":
    main()

File: Fundamentals_of_Software_Design/1_Basic_skills/Счастливые билеты (продвинутая версия)/main.c
#include <stdio.h>
#include <stdlib.h>

int is_lucky_ticket(char* ticket) {
    int sum1 = 0, sum2 = 0;

    // Считаем сумму первых 16 цифр
    for (int i = 0; i < 16; i++) {
        sum1 += ticket[i] - '0';  // Преобразуем символ в цифру
    }

    // Считаем сумму последних 16 цифр
    for (int i = 16; i < 32; i++) {
        sum2 += ticket[i] - '0';  // Преобразуем символ в цифру
    }

    return sum1 == sum2;  // Если суммы равны, билет счастливый
}

int main() {
    int total_tickets = 1000000;  // Количество билетов для проверки
    int lucky_count = 0;  // Количество счастливых билетов

    // Генерируем случайные билеты
    for (int i = 0; i < total_tickets; i++) {
        // Генерируем случайный 32-значный номер билета
        char ticket[33];  // Массив для хранения билета
        for (int j = 0; j < 32; j++) {
            ticket[j] = '0' + (rand() % 10);  // Случайная цифра от 0 до 9
        }
        ticket[32] = '\0';  // Завершающий ноль для строки

        // Проверяем, является ли билет счастливым
        if (is_lucky_ticket(ticket)) {
            lucky_count++;
        }
    }

    // Вычисляем долю счастливых билетов
    double lucky_percentage = (double)lucky_count / total_tickets * 100;

    printf("Доля счастливых билетов: %.6f%%\n", lucky_percentage);

    return 0;
}
